MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/module/http/HttpRequestImpl.java

```diff
@@ -155,27 +112,54 @@ public void onResponse(@NonNull Call call, @NonNull Response response) {
         response.header("Retry-After"),
         response.header("x-rate-limit-reset"),
         body);
+
+    } catch (Exception exception) {
+      handleFailure(httpRequest, call, exception);
     }
+  }
 
-    private void handleFailure(@Nullable Call call, Exception e) {
-      String errorMessage = e.getMessage() != null ? e.getMessage() : "Error processing the request";
-      int type = getFailureType(e);
+  private void handleFailure(@NonNull HttpResponder httpRequest, @Nullable Call call, Exception e) {
+    String errorMessage = e.getMessage() != null ? e.getMessage() : "Error processing the request";
+    int type = getFailureType(e);
 
-      if (HttpLogger.logEnabled && call != null && call.request() != null) {
-        String requestUrl = call.request().url().toString();
-        HttpLogger.logFailure(type, errorMessage, requestUrl);
-      }
-      httpRequest.handleFailure(type, errorMessage);
+    if (HttpLogger.logEnabled && call != null) {
+      String requestUrl = call.request().url().toString();
+      HttpLogger.logFailure(type, errorMessage, requestUrl);
+    }
+    httpRequest.handleFailure(type, errorMessage);
+  }
+
+  private int getFailureType(Exception e) {
+    if ((e instanceof UnknownHostException) || (e instanceof SocketException)
+      || (e instanceof ProtocolException) || (e instanceof SSLException)) {
+      return CONNECTION_ERROR;
+    } else if ((e instanceof InterruptedIOException)) {
+      return TEMPORARY_ERROR;
     }
+    return PERMANENT_ERROR;
+  }
 
-    private int getFailureType(Exception e) {
-      if ((e instanceof NoRouteToHostException) || (e instanceof UnknownHostException) || (e instanceof SocketException)
-        || (e instanceof ProtocolException) || (e instanceof SSLException)) {
-        return CONNECTION_ERROR;
-      } else if ((e instanceof InterruptedIOException)) {
-        return TEMPORARY_ERROR;
-      }
-      return PERMANENT_ERROR;
+  @Override
+  public void cancelRequest() {
```

Comment:
Which thread manages and cancels the requests? Is it the same thread that creates the requests? Can it lead to deadlocks or prevent actual cancellation?

Reply:
I'm observing the following in both this as old implementation. If you pan the map to a new area that requires loading tiles:

First I see:
```
[HTTP] Request was successful (code = 200). https://api.mapbox.com/v4/mapbox.mapbox-streets-v8,mapbox.mapbox-terrain-v2/5/18/15.vector.pbf?access_token=pk.eyJ1IjoidG9icnVuIiwiYSI6ImNpajVlajR0cjAwNjN2NmtyY204eWw2eG0ifQ.x3_WEoExNW5Qyv9T3Vj7Mw&sku=100k3qaidfqebac703879074193acfa72f8a283b183
```

Followed by:
```
CancelRequest for call: https://api.mapbox.com/v4/mapbox.mapbox-streets-v8,mapbox.mapbox-terrain-v2/5/18/15.vector.pbf?access_token=pk.eyJ1IjoidG9icnVuIiwiYSI6ImNpajVlajR0cjAwNjN2NmtyY204eWw2eG0ifQ.x3_WEoExNW5Qyv9T3Vj7Mw&sku=100k3qaidfqebac703879074193acfa72f8a283b183
```

I initially thought this wasn't expected behavior but we do execute cancel as part of the destructor of HTTPRequest. When I optimise the code to only excute cancel when we haven't received a response. I'm seeing that the old implementation does occasionally cancel a request while this PR doesn't.

Reply:
I will follow up with a patch for mapbox-gl-native (for http_request.ccp) but will close this PR as not actionable.